[GAWB-1826] Give users access to JES metadata for calls inside a workflow

rtitle opened the pull request:
JIRA:
https://broadinstitute.atlassian.net/browse/GAWB-1826

Related PRs:
https://github.com/broadinstitute/firecloud-orchestration/pull/445
https://github.com/broadinstitute/firecloud-ui/pull/785

Removed `'/api/admin/genomics/operations/{jobId}` and added `/api/workspaces/genomics/operations/{jobId}` (without admin privileges).

Cleaned up some imports using the IDE.

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [x] **Submitter**: Make sure Swagger is updated if API changes
  - [x] **...and Orchestration's Swagger too!**
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] Tell your tech lead (TL) that the PR exists if they want to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [x] **TL** sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Squash commits and merge to develop
- [x] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

rtitle committed:
[GAWB-1826] Give users access to JES metadata for calls inside a workflow

helgridly commented:
This allows you to read metadata for operation IDs that aren't yours. Should it?

rtitle commented:
Whoops did I miss some permissions stuff? Should it be doing some workspace access checks (maybe `WorkspaceAccessLevels.Read` or higher)?

I guess that would also mean passing the workspace info in the API call..

helgridly commented:
a) I misread the PR initially, so that's not true (but there might be a wrinkle regardless); and
b) I should have done a better job explaining the background here -- sorry.

So operation IDs are the JES IDs. They're not associated with workspaces, because JES has no concept of workspaces - it just runs jobs. opIDs _are_ associated with Google projects, so if you ask for the opID metadata as a user, JES will check you're in the relevant project and 40x as appropriate. The wrinkle is that I'm not sure which scopes are required for this operation, so we should find that out, since we'll need to ask the user for them if they haven't already authorized us.

jmthibault79 commented:
I am also unsure about whether we should care if the user can access Operation ID Metadata outside of the Workspace in question.  Let's ask the PO.

If we don't care: 👍 for this PR as-is.

If we do, then requiring `Read`+ access sounds right to me, and I'd make the URL `/api/workspaces/{ns}/{name}/genomics/operations/{jobId}`

helgridly commented:
Actually, I lied _again_ - `GoogleDAO.getGenomicsOperation` uses rawls' genomics-credential for the query. So this is broken -- we should be using the user's.

helgridly commented:
Which means we need to:

a) find out where we define the list of scopes the user needs, and make sure `GenomicsScopes.GENOMICS` is in that list;
b) pass in the user's credential (like in [`listBillingAccounts`](https://github.com/broadinstitute/rawls/blob/develop/core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala#L659)) and execute the Google request as them.

helgridly commented:
[Here](https://github.com/broadinstitute/firecloud-orchestration/blob/fa61350f915debf800519bc6c7bd80a85530853c/src/main/scala/org/broadinstitute/dsde/firecloud/service/OAuthService.scala#L58) is where we define the scopes we want from the user. Thanks Joel.

rtitle commented:
Ah wow, thanks guys. Will work on those changes.

rtitle committed:
Update GoogleServicesDAO.getGenomicsOperation to use the user account (with appropriate genomics scopes) instead of a service account

rtitle commented:
Pushed commits to make GoogleServicesDAO.getGenomicsOperation use the user's credential (with the appropriate scopes) instead of a service account. I now can't access my test operation with my user id. :)

```
Response Body
{
  "statusCode": 403,
  "source": "rawls",
  "causes": [],
  "stackTrace": [],
  "message": "Permission denied; need genomics.operations.get permission for operation EO_hv-6-KxibjYjL0OCrmCQgw7vetLsXKg9wcm9kdWN0aW9uUXVldWU"
}
Response Code
403
```

I posted the question about whether we also need to care about workspace permissions to the JIRA: https://broadinstitute.atlassian.net/browse/GAWB-1826?focusedCommentId=91968&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-91968

rtitle committed:
Update GenomicsService to propagate Google status and error message to the caller

rtitle committed:
Oops, fix compilation error

helgridly commented:
Ugh. I had assumed that if you have access to a workspace, you have access to the Google Project. But we only give you access to the bucket!

So: the API should live at `/api/workspaces/<namespace>/<name>/operationID/<opID>`, and should check your workspace permissions. Then it should use the Rawls service account (which has genomics scopes) to actually query JES.

Lesson learned: stop assuming things. Mea culpa.

rtitle commented:
Ok

rtitle committed:
Update workspaces genomics endpoint to validate the workspace permissions and use a service account for the Google Genomics API

rtitle committed:
Fix complilation error due to import cleanup

rtitle committed:
Fix unit test failure

rtitle committed:
Remove google genomics scope from swagger

rtitle commented on core/src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala:
```diff
@@ -28,11 +27,10 @@ import spray.can.Http
-
-import scala.concurrent.ExecutionContext.Implicits.global
 import scala.collection.JavaConversions._
+import scala.concurrent.ExecutionContext.Implicits.global
```
While cleaning up imports I noticed that we're using the global execution context. I was worried that it was also being used for slick blocking calls, but thankfully I see we're using a separate thread pool for slick.

This is probably fine for our case, but just fyi another way to do it is to define dispatchers within akka and reference them using the `ActorSystem`. This allows you to define & tune all execution contexts from within akka.

http://doc.akka.io/docs/akka/current/scala/dispatchers.html

geoffjentry replied:
Also IIRC Rawls uses the `blocking` construct so if you ever *do* do what @rtitle suggests note that it only does anyting on the global ec

helgridly replied:
We don't use `blocking` often, I don't think.

rtitle replied:
I see use of `blocking` on Google HTTP calls, and a separate thread pool for slick, which is good. :)

rtitle commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GoogleServicesDAO.scala:
```diff
@@ -127,7 +128,7 @@ abstract class GoogleServicesDAO(groupsPrefix: String) extends ErrorReportable {
   def revokeToken(rawlsUserRef: RawlsUserRef): Future[Unit]
 
-  def getGenomicsOperation(jobId: String): Future[Option[JsObject]]
+  def getGenomicsOperation(jobId: String): Future[JsObject]
```
Just using Future.failed to represent the exception case instead of Option.

rtitle commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -753,21 +749,19 @@ class HttpGoogleServicesDAO(
-      case t: HttpResponseException if t.getStatusCode == StatusCodes.NotFound.intValue => None
+      executeGoogleRequest(operationRequest).toPrettyString.parseJson.asJsObject
+    }) { case e: GoogleJsonResponseException =>
+      val statusCode = StatusCodes.getForKey(e.getStatusCode).getOrElse(StatusCodes.InternalServerError)
```
This propagates the status code + message from the Google HTTP call to the end user.

helgridly replied:
What if it's not a `GoogleJsonResponseException`?

rtitle replied:
This is a `PartialFunction`, so the exception will complete the Future and the API will return a 500 (I think).

helgridly replied:
This is a significant change to the previous behaviour. Before, non-404s would potentially get retried because they'd fall through and get caught in `GoogleUtilities.when500orGoogleError`. Now we're just passing the Google error straight up the chain, wrapped in a `RawlsExceptionWithErrorReport`, so it bypasses our retry mechanism.

I'd guess this is probably why the `Option` was there -- to distinguish between a 404-failure (which just means we couldn't find your job, turns into `None`) and a retry-able failure (might succeed next time, or otherwise return a properly-failed `Future` if it's legit buggered).

rtitle replied:
I think the retries still work as before - the `recover` partial function only gets invoked if the operation still fails after all retries are done.

W.r.t. to the response status, I think it's better behavior. :) The Rawls response now gives information about what went wrong in Google.

### Old behavior:
- Case 1
  - Google returns 404
  - This method "catches" the 404 and returns None
  - Rawls endpoint returns 404 anyway: 
```case None => RequestComplete(StatusCodes.NotFound, s"jobId $jobId not found.")```
- Case 2
  - Google returns 403 (or anything else)
  - Exception is not "caught" by this method
  - Rawls endpoint returns 500

### New behavior:
- Google returns an error
- Google status + message is wrapped in `RawlsExceptionWithErrorReport`
- Rawls API returns the Google status + message

rtitle replied:
Crap, I understood `retryWithRecoverWhen500orGoogleError` wrong.  Looking more...

rtitle commented:
@jmthibault79 @helgridly re-review when you have a sec? I think it's correct now :)

jmthibault79 commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -1804,6 +1798,15 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+    dataSource.inTransaction { dataAccess =>
+      withWorkspaceContextAndPermissions(workspaceName, WorkspaceAccessLevels.Read, dataAccess) { workspaceContext =>
+        val genomicsServiceRef = context.actorOf(GenomicsService.props(genomicsServiceConstructor, userInfo))
+        DBIO.from((genomicsServiceRef ? GenomicsService.GetOperation(jobId)).mapTo[PerRequestMessage])
```
Since this is an external call, we should move this outside the transaction.

rtitle replied:
Makes sense.

rtitle committed:
Code review comments:
- Don't change retry behavior for genomics call
- Call the genomics API outside of the DB transaction

rtitle commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -753,21 +751,25 @@ class HttpGoogleServicesDAO(
+    // This uses the "applicative builder" syntax from cats to combine 2 predicates
+    // (when500orGoogleError and whenGoogle404) using a binary operator (e.g.
+    // (Boolean, Boolean) => Boolean), and return a resulting predicate.
+    val predicate = (when500orGoogleError |@| whenGoogle404).map((a, b) => a && !b)
```
I hope you guys aren't too annoyed at me using cats again. :)

In this case I think it's cleaner: combining predicates using `|@|` and passing the combined predicate to the retry loop; rather than hard-coding the predicate, jamming a `recover` on the Future, and using `Option` to represent a "special" recoverable value to sort of override the predicate.

helgridly replied:
I don't have a problem with the `|@|`, but I'd rename `predicate` to something a bit more descriptive.

I still think this is less clear to users. You're treating 404 as a failure when it's an expected return status for the use case. The user will get whatever Google responds with, and it'll be caught by `PerRequest.logResponse` and logged as a warning. I don't think either of these are desirable; there's a lot to be said for transforming expected errors into something user-readable.

rtitle replied:
I'll rename `predicate`.

Not sure I understand entirely. The previous logic caught Google 404's and returned Rawls 404's (with a similar error message). See https://github.com/broadinstitute/rawls/pull/694/files#diff-16a4ec11905001338e194fb84e2fad72L43. The new behavior for the 404 case is essentially the same:
```
https://local.broadinstitute.org:20443/api/workspaces/broad-dsde-dev/rtitle-1826-test/genomics/operations/foo

{
  "statusCode": 404,
  "source": "rawls",
  "causes": [],
  "stackTrace": [],
  "message": "Job not found: \"foo\""
}
```
The difference with the new code is, if Google returns an error besides 404, before the Rawls API would return 500 with the actual Google error buried somewhere in the stack trace. Now, the Google response status and message are used in the Rawls response. The status + message are displayed to the user in a modal dialog, so I thought it would be more useful to propagate the Google error instead of just showing a generic 500.

But if you feel this is not ideal I can change it.

rtitle commented on google/src/main/scala/org/broadinstitute/dsde/rawls/google/GoogleUtilities.scala:
```diff
@@ -36,11 +36,22 @@ trait GoogleUtilities extends LazyLogging with Retry {
+    }
+  }
+
+  protected def retryWhen[T](pred: Predicate[Throwable])(op: () => T): Future[T] = {
```
I feel like all these `op: () => T` (which is a Function1[Unit, T]) could be written using by-name parameters instead: `op: => T`. But I'm not sure if there is a good reason for this, so I left it as is.

rtitle replied:
I'm also thinking of adding a spec for the `Retry` logic, since the logic tripped me up and I don't see any existing tests for it.

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -753,21 +751,25 @@ class HttpGoogleServicesDAO(
+    val predicate = (when500orGoogleError |@| whenGoogle404).map((a, b) => a && !b)
+
+    retryWhen(predicate) { () =>
+      executeGoogleRequest(operationRequest).toPrettyString.parseJson.asJsObject
```
You've lost the helpful comment that was above this line.

rtitle replied:
whoops, thanks

rtitle committed:
Restore previous behavior w.r.t. Google genomics error cases

rtitle committed:
Added a RetrySpec

rtitle commented:
@helgridly @jmthibault79 I think this is ready for another pass.  Changes since the last review:
1. Reverted previous behavior of `HttpGoogleServicesDAO.getGenomicsOperation` w.r.t. retries and response codes
2. Run Google genomics call outside of the DB transaction
3. Added a `RetrySpec` to increase code coverage by 0.0000001% :)

rtitle committed:
Beef up RetrySpec a bit

helgridly commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -1804,6 +1798,21 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+    // Next call GenomicsService, which actually makes the Google Genomics API call.
+    .flatMap { _ =>
+      val genomicsServiceRef = context.actorOf(GenomicsService.props(genomicsServiceConstructor, userInfo))
+      (genomicsServiceRef ? GenomicsService.GetOperation(jobId)).mapTo[PerRequestMessage]
```
I had to look this up: `mapTo[PerRequestMessage]` is a less sucky version of `asInstanceOf[Future[PerRequestMessage]]`.

rtitle replied:
Yes! Always use `mapTo` instead of `asInstanceOf` to safely cast a Future to an expected type.
(It's often needed with Akka because actors are untyped 😢 )

geoffjentry replied:
[Not for long](http://blog.akka.io/typed/2017/05/05/typed-intro)

rtitle replied:
What! How have I not seen this

geoffjentry replied:
Well this blog post is only a week or two old, althogh they've been pushing out sequels at a good clip. I've been taking an "I'll believe it when I see it" approach as typed akka has been Coming Soon (including here and gone already) for so many years, but it looks like they're serious this time.

helgridly commented on util/src/test/scala/org/broadinstitute/dsde/rawls/util/RetrySpec.scala:
```diff
@@ -0,0 +1,179 @@
+import org.scalatest.mock.MockitoSugar
+import org.scalatest.time.{Minutes, Seconds, Span}
+import org.scalatest.{BeforeAndAfterAll, FlatSpecLike, Matchers}
+import org.slf4j.{Logger => Underlying}
```
Why did you rename this? I'm confused every time I see it.

rtitle replied:
To deal with name conflicts between `org.slf4j.Logger` and `com.typesafe.scalalogging.Logger`. `Retry` mixes in `LazyLogging` which uses `com.typesafe.scalalogging.Logger`. But it turns out `com.typesafe.scalalogging.Logger` is not mockable because it's a `final` class, so I'm mocking the underlying `org.slf4j.Logger` instead and passing it to the `com.typesafe.scalalogging.Logger` constructor:
https://github.com/typesafehub/scalalogging/blob/master/scalalogging-slf4j/src/main/scala/com/typesafe/scalalogging/slf4j/Logger.scala#L38

I agree it's confusing, maybe I'll rename the type to `SLF4JLogger` for clarity.

helgridly commented on util/src/test/scala/org/broadinstitute/dsde/rawls/util/RetrySpec.scala:
```diff
@@ -0,0 +1,179 @@
+class RetrySpec extends TestKit(ActorSystem("MySpec")) with FlatSpecLike with BeforeAndAfterAll with Matchers with MockitoSugar with ScalaFutures {
+  import system.dispatcher
+
+  implicit override val patienceConfig = PatienceConfig(timeout = scaled(Span(10, Seconds)))
```
I'm assuming this is some Mockito magic. Could you comment it?

rtitle replied:
Actually not Mockito, it's ScalaTest Futures:
http://doc.scalatest.org/2.2.4/index.html#org.scalatest.concurrent.Futures

Basically it configures how long the calls to `whenReady` will wait. I'll comment it.

helgridly commented on util/src/test/scala/org/broadinstitute/dsde/rawls/util/RetrySpec.scala:
```diff
@@ -0,0 +1,179 @@
+      ex shouldBe an [Exception]
+      ex should have message "test exception"
+      testable.invocationCount should equal(4)
+      verify(testable.underlying, times(4)).info(anyString, any[Throwable])
```
This is `|+|` all over again :) Could you slap a comment on here explaining what's going on? It took me a while to figure out the meaning, and I _still_ find `verify().info()` bizarre phrasing (shouldn't the thing you're testing against be inside the verify function?).

rtitle replied:
Yeah Mockito takes some getting used to. It's a pretty powerful testing tool though. `verify` takes in a mock and an expected number of times, then you can call methods on the returned mock with argument matchers to verify that the mock was called with the right arguments.

I'll add some comments. :)

helgridly commented on util/src/test/scala/org/broadinstitute/dsde/rawls/util/RetrySpec.scala:
```diff
@@ -0,0 +1,179 @@
+
+}
+
+class TestRetry(val system: ActorSystem, val underlying: Underlying) extends Retry with LazyLogging {
```
Please also rename `underlying` here.

rtitle replied:
👍

helgridly commented on util/src/test/scala/org/broadinstitute/dsde/rawls/util/RetrySpec.scala:
```diff
@@ -0,0 +1,179 @@
+  it should "retry until a success" in {
+    val testable = new TestRetry(system, setUpMockLogger)
+
+    val result = testable.retryUntilSuccessOrTimeout()(100 milliseconds, 1 day)(() => testable.failureNTimes(10))
```
Can we make this a _bit_ less than a day? Surely a minute is enough.

rtitle replied:
Sure, 1 day is excessive. Though the idea here is to _not_ hit the timeout -- `whenReady` will give up anyway after 10 seconds.

helgridly reviewed (approved):
Some nitpicks and comment-requests, but go ahead and merge this after they're done. 👍

rtitle commented:
Thanks! Fixing nitpicks, then will merge.

jmthibault79 commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -1804,6 +1798,21 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+      }
+    }
+    // Next call GenomicsService, which actually makes the Google Genomics API call.
+    .flatMap { _ =>
```
I usually use `andThen` in this case.  Is that just a style difference or is this way better?

rtitle replied:
Actually `andThen` wouldn't work in this case because it returns the result of the _first_ future. Here I want to return the result of the second future, but only if the first future is successful. So I think `flatMap` is appropriate.

jmthibault79 replied:
hmmmmmmm I hope I have not been using `andThen` incorrectly.

Thanks.

rtitle replied:
It's probably fine, or else it wouldn't compile :)

jmthibault79 commented on google/src/main/scala/org/broadinstitute/dsde/rawls/google/GoogleUtilities.scala:
```diff
@@ -36,11 +36,15 @@ trait GoogleUtilities extends LazyLogging with Retry {
   }
 
-  protected def retryWhen500orGoogleError[T](op: () => T) = {
+  protected def retryWhen[T](pred: Predicate[Throwable])(op: () => T): Future[T] = {
```
This doesn't seem to be referenced anywhere.  Did you intend to compose the other `retryWhen`s using this?

rtitle replied:
Ah good catch, it's not being used anywhere. I was using it earlier but refactored code such that it's no longer needed. I can just remove it.

jmthibault79 commented on util/src/main/scala/org/broadinstitute/dsde/rawls/util/Retry.scala:
```diff
@@ -15,13 +15,15 @@ trait Retry {
   this: LazyLogging =>
   val system: ActorSystem
 
+  type Predicate[A] = A => Boolean
```
I like this.  Would you mind bringing `def always` up here and defining it in terms of `Predicate`?

rtitle replied:
👍

rtitle replied:
Done

rtitle committed:
Code review feedback: mostly more comments cleanup

jmthibault79 commented:
👍 but I think we are waiting on QA's approval for merges to dev at the moment
